### PR TITLE
Build NuPIC.cpp as shared library.

### PR DESCRIPTION
**Pull request by marty1885**
Hi, I found that NuPIC.cpp can be built and ran perfectly as shared library. This should reduce the build time and resource  for applications using NuPIC.cpp and enable a more liberality use of libraries in NuPIC.cpp.

I have only tested it on Linux as I don't have a Windows box that is capable of doing it. More test is needed.

### CODE REVIEW

**Comment by marty1885** [comment]
Weird. It works for me.

https://gist.github.com/marty1885/02eef38ec2c754576bc5ff3181c066d3

**Comment by breznak** [comment]
> Perhaps a better way would be to build both the static library and a
dynamic library. So, lets add a new target with the same arguments as our
static library except we mark it as SHARED and call it something like
nupic_core_shared.so

I like the approach of offering both. We can add configure option to choose
which get built.

If windows is too much of a trouble, this might be only for Linux

And thanks for the PR

-- 
Marek Otahal :o)

**Comment by dkeeney** [comment]
I have a lot on my plate at the moment but perhaps later this week I can take a look at this in more detail and see if I can come up with a solution.  I would think that a Windows C++ only developer  would also want a Dll library. So it would be good if we could accommodate both types of libraries on all platforms.

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -62,13 +62,7 @@ configure_file(
   
 
 ############ nupic_core Library ######################################
-# Setup nupic_core static library, consisting of our own sources and yaml libraries;
-#
-# this is an intermediate archive that will be merged with external static
-# libraries in a subsequent step to produce the nupic_core "combined" static
-# library.
-#
-set(src_lib_static_nupiccore_solo nupiccore_solo)
```
Comment (breznak) [inline src/CMakeLists.txt]: @dkeeney I remembered there was a reason we chose static. Couldn’t all external and ours be made shared?

Comment (dkeeney) [reply src/CMakeLists.txt]: There are times when building a C++ app where you really want a static lib.  I think we need to provide both.

**Comment by breznak** [comment]
@marty1885 if you have time, revisit the feedback, please

**Comment by dkeeney** [comment]
Ok, now that my previous project is finished I can take a look at creating a shared library for C++.
Since the static library is used to create the python extensions we need to keep that but we can also create a dynamic library from the same sources.

For Linux, no problem.  But for a windows DLL every class that is to be exposed must be declared as being exported.  So we have to modify every .h file.

I am adding the following to Log.hpp which is included by everything.
```
#if defined (WIN32) && defined (BUILD_SHARED_LIBS)
#if defined (_MSC_VER)
#pragma warning(disable: 4251)
#endif
   #if defined(DllLib_EXPORT)
     #define  DllLIB_EXPORT __declspec(dllexport)
   #else
     #define  DLLLIB_EXPORT __declspec(dllimport)
   #endif
#else
  #define DllLIB_EXPORT
#endif
```
Then for every class, we add DLLLIB_EXPORT as follows:
`   class DLLLIB_EXPORT Anomaly { `

For non-windows environments, DLLLIB_EXPORT  evaluates to nothing and has no effect.

**Comment by dkeeney** [comment]
As of CMake 3.4 there is a new CMake feature -DCMAKE_WINDOWS_EXPORT_ALL_SYMBOLS=TRUE
see https://blog.kitware.com/create-dlls-on-windows-without-declspec-using-new-cmake-export-all-feature/
This will do the dllexport/dllimport thing for me as long as we don't export global variables.  Should not do that anyway.  So I will give this a try.

**Comment by breznak** [comment]
> This will do the dllexport/dllimport thing for me as long as we don't export global variables. Should not do that anyway. So I will give this a try. 

I like that much better than hacking Log.hpp to include the IFDEFs for Win.

**Comment by dkeeney** [comment]
yes, me too.  Keep in mind that this will NOT work with exported global data (should not have any anyway).
There is some question if it will work with templated classes.  I have to experiment.

**Comment by dkeeney** [comment]
Well, I cannot get CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS to work.  There is an error but I cannot figure out what the error actually is.  My guess is that it gets very confused with the templated classes.  

So, I guess the only alternative is to modify all sources and stick in a macro on each exported class and each exported function that is not in a class.  This does have the advantage that we can control what is exported and we can hide all of the internal classes and functions.  I will take a first stab a what those might be and you two can add or remove as you see fit.

There will be a macro defined in Log.hpp (everything includes this) which will select dllexport or dllimport depending on how it is compiled.  for Linux and OSx it does nothing.

**Comment by dkeeney** [comment]
After struggling around in DLL hell for a while I now realize why they call it that.  Making a shared library for Linux or OSx does not seem to be a problem.  But making a DLL for windows definitely is a problem because the mechanism is different.  There are some significant rules that must be followed to get it to compile and not crash if it does compile.
- All exported symbols must be declared.  There is a way to do this that is not too hard.
- No globals in the DLL.  Those are not good practice anyway so no problem.
- No way to initialize a static member variable (unless its a const).  The DLL thinks of these that same as a global. We have several singleton classes in our library so its a problem.  Basically, a DLL cannot contain anything that maintains state outside an instance.
- No templates.  Well, yes you can have templates but the .cpp side of the class must explicitly implement all instances that a template variable can take on.  Not just a class but functions too. For example, 
```
    template<typename T>
    void setDense( const vector<T> &value ) {  ... some implementation... }
```
If this is in the hpp, no problem.   The compiler builds the implementations needed.  But if you put this on the .cpp side which would mean it is exported from the DLL then the function setDense(const vector<T> &value ) would need to be explicitly implemented for each value of T that is expected.  
```
setDense(const vector<Byte> &value ) {  ... implementation using Byte... }
setDense(const vector<UInt> &value ), {  ... implementation using UInt... }
etc.
```
- No exceptions can cross the DLL boundary.   Not totally positive that this is still the case but you do have to play games to get exceptions to work.

**Comment by breznak** [comment]
> in DLL hell for a while I now realize why they call it that. Making a shared library for Linux or OSx does not seem to be a problem. But making a DLL for windows definitely is a problem  

with all the Win requirements you describe below, I'd say it adds a lot of burden for a small (how much needed?) feature. 

Let's make shared lib available only for Linux/OSX, and if there's a huge demand for that on Win, we can revisit this issue. 

> Basically, a DLL cannot contain anything that maintains state outside an instance.

this would be a limitation

> If [the template`<T>` definition] this is in the hpp, no problem.

we should be good. I don't know if it's a cpp language limitation, but we have all templated definitions+implementations in hpp only. (does not work: 
```
hpp: void foo<T>(T a); 
cpp: void foo<T>(T a) { print a;}
```

**Comment by ctrl-z-9000-times** [comment]
>  No way to initialize a static member variable (unless its a const).  The DLL thinks of these that same as a global. We have several singleton classes in our library so its a problem.  Basically, a DLL cannot contain anything that maintains state outside an instance.

This is going to be a problem for the random number generator, which keeps a global RNG to use for when the caller doesn't specify a seed (defaults to seed == 0).  This RNG also has a global flag which tracks whether it's been initialized.  We need a global variable for the RNG, so that in debug mode two RNG's initialized with seed == 0 end up with different seeds.

I think it should be possible to use `DllMain` to initialize global variables.

**Comment by dkeeney** [comment]
I think the rules for DLL given previously are too pessimistic.  I am going to try this again.  There must be a way to make this work.

**Comment by dkeeney** [comment]
Singletons: Unlike in Linux; in Windows a DLL loads only one copy of the library into memory and that code is shared by multiple processes at the same time.  To do that each connecting process maps the DLL to its address space (The reason for position independent code).  This is also the reason we cannot set a static variable except from within a function in the DLL.  But for singletons it must have an already known value like a null pointer so you know you are the first thread to access it.  That is what DLLMain is for. It is like a separate process that runs when  the DLL is loaded and will initialize stuff before any process can connect to it.   So, it is possible to have our singleton classes.  I just need to setup this DLLMain.

**Comment by dkeeney** [comment]
Exceptions:  We are using std:: C++ exceptions for error reporting.  The problem is that if one of these exceptions cross the DLL boundary, it cannot be caught and the program crashes.  We could use SEH exceptions which do cross the boundary but the mechanism is different. One approach is to catch all C++ exceptions before they cross the boundary  and re-throw them as SEH exceptions.  Then on the application program side, we write a magic function called __set_se_translator_ to catch the SEH exceptions and re-throw them as standard C++ exceptions.

This is ugly.  But we could write wrapper code that does this. It would end up looking very similar to the code we have for the .py bindings.  After all, the .pyd for windows actually is a .dll.  So, maybe a bindings/dll directory would be the place to put all of that code.

**Comment by dkeeney** [comment]
Templates:  The calling side of a templated class or function has the template variable specified.  All we need to do is add something in our dll wrapper some code that can resolve to whatever we might be called with.  We cannot do this with unknown types but we can do this with specializations.

In the static library's header we might have something like this:
```
template<typename T>
void TestPointModule_Check(string name, T* value);
```
In our DLL wrapper .hpp , add one of these for each type it can resolve to.
```
template <>
DLL_Export void TestPointModule_Check<AParticularType>
                    (string name, AParticularType* value);
```
In the wrapper .cpp we add something to call into our static library.
```
template <>
DLL_Export void TestPointModule_Check<AParticularType>
                    (string name, AParticularType* value) {
    virtual_tpm->CheckTestPoint<AParticularType>(name, value);  
}
```

**Comment by dkeeney** [comment]
To summarize, we need to write some macros for the DLL bindings that are similar to PyBind11.  In fact, perhaps I can de-construct their macos to figure out how to do this since they solved the same problems.
- Something to initialize static class variables using DLLMain.
- Something to capture standard C++ exceptions and pass them on.
- Something to handle templates.  Although this might just require manually writing code.

Needless to say, this is not a simple task.
What I am going to do is implement shared libraries for Linux and OSx and leave MSVC for another PR at a lower priority.

**Comment by breznak** [comment]
> This is ugly. But we could write wrapper code that does this. 

This is starting to be really complicated. Is there really need/interest for the Win DLL part? So far we had only static linking and it was ok. 

I'll be happy for the code, if you want to work on it. But it should not place (too much) restrictions on how we code. Or too complicated wrappers. What do you think?

**Comment by dkeeney** [comment]
I am going to do the .so library but not the .dll library with this PR.
The .so library has no such restrictions.

**Comment by dkeeney** [comment]
@marty1885 , Thank you for your PR.  I ended up implementing shared library only for Linux and OSx.  You should find it in the htm-community:master repo as soon as PR #261 is approved.

Windows DLL became a rabbit hole. It would have required significant changes to the source code just to have symbols to export so I did not implement the DLLs.  If there seems to be significant interest in the community for a DLL I will re-visit this project.

**Pull request closed without merging**